### ISSUE TITLE
Complete method self param declarations

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#9522]
Does this issue consider completing `self` like here:

```rust
impl Foo {
    fn bar(<|>) {
    }
}
```

It'd be cool if this completed `self`, `&self`, `&mut self`, `mut self` and `self: Arc<Self>`

_Originally posted by @jhgg in https://github.com/rust-analyzer/rust-analyzer/issues/9311#issuecomment-875249115_

A simple implementation has been done in #9548 but this only works for empty parameterlists which is not too useful as typing `&` already doesn't offer any completions.

**Comment by flodiebold** [comment]
I'd propose we complete `self`, `&self`, `mut self` and `&mut self`. I don't know about special-casing `Arc<Self>`.

**Comment by jhgg** [comment]
> I'd propose we complete `self`, `&self`, `mut self` and `&mut self`. I don't know about special-casing `Arc<Self>`.

Fine with me. It would be interesting if we could offer `Arc<Self>` as a completion in this position though, as it's a fairly common pattern to do... 🤔 

```
impl Foo {
   fn bar(self: <|>) {
   }
}
```

**Comment by lhvy** [comment]
Is this issue resolved with #9548 merged?

**Comment by Veykril** [comment]
It isn't, that PR only added a very basic form of this. Basically it only works for functions with no parameters in them, see the comments after the merge in https://github.com/rust-analyzer/rust-analyzer/pull/9548#issuecomment-877725649 as well as https://github.com/rust-analyzer/rust-analyzer/pull/9575
